# Required libraries :

In [1]:
import pandas as pd
import numpy as np
import glob, os

<mark>A noter j'ai supprimé les données concernant 2009 car n'a pas la même forme que les autres ( juste pour lire tout en même temps)</mark>

#### Les données sur les caracteristiques :

In [2]:
carac_2005=pd.read_csv('dataset/caracteristiques/caracteristiques_2005.csv',encoding='latin-1')
carac_2005.shape

(87026, 16)

In [3]:
carac_2005.head(2)

Num_Acc  an  mois  jour  hrmn  lum  agg  int  atm  col   com  \
0  200500000001   5     1    12  1900    3    2    1    1    3  11.0   
1  200500000002   5     1    21  1600    1    2    1    1    1  51.0   

            adr gps        lat      long  dep  
0         CD41B   M  5051500.0  294400.0  590  
1  rue de Lille   M  5053700.0  280200.0  590

In [4]:
path =r'dataset/caracteristiques' # use your path
allFiles = glob.glob(path + "/*.csv")
frame = pd.DataFrame()
list_ = []
for file_ in allFiles:
    df = pd.read_csv(file_,encoding='latin-1')
    list_.append(df)
caracteristiques = pd.concat(list_)

In [5]:
caracteristiques.shape

(765576, 16)

##### Les données sur les usagers :

In [6]:
usager_2005=pd.read_csv('dataset/usagers/usagers_2005.csv',encoding='latin-1')
usager_2005.shape

(197498, 12)

In [7]:
usager_2005.head(2)

Num_Acc  place  catu  grav  sexe  trajet  secu  locp  actp  etatp  \
0  200500000001      1     1     4     1       1    11     0     0      0   
1  200500000001      1     1     3     2       3    11     0     0      0   

   an_nais num_veh  
0   1976.0     A01  
1   1968.0     B02

In [8]:
path =r'dataset/usagers' # use your path
allFiles = glob.glob(path + "/*.csv")
frame = pd.DataFrame()
list_ = []
for file_ in allFiles:
    df = pd.read_csv(file_,encoding='latin-1')
    list_.append(df)
usagers = pd.concat(list_)

In [9]:
usagers.shape

(1710043, 12)

In [10]:
usagers.head(2)

Num_Acc  place  catu  grav  sexe  trajet  secu  locp  actp  etatp  \
0  200500000001    1.0     1     4     1     1.0  11.0   0.0   0.0    0.0   
1  200500000001    1.0     1     3     2     3.0  11.0   0.0   0.0    0.0   

   an_nais num_veh  
0   1976.0     A01  
1   1968.0     B02

#### Les données sur les lieux :

In [11]:
lieu_2005=df1=pd.read_csv('dataset/lieux/lieux_2005.csv',encoding='latin-1')
lieu_2005.shape

(87026, 18)

In [12]:
lieu_2005.head(2)

Num_Acc  catr  voie   v1   v2  circ  nbv   pr    pr1  vosp  prof  \
0  200500000001   3.0  41.0  0.0    B     2    2  1.0  430.0     0     1   
1  200500000002   2.0  41.0  0.0  NaN     0    2  0.0    0.0     1     1   

   plan  lartpc  larrout  surf  infra  situ  env1  
0     1       0       63     1      0     1     0  
1     1       0      100     1      0     5     0

In [13]:
path =r'dataset/lieux' # use your path
allFiles = glob.glob(path + "/*.csv")
frame = pd.DataFrame()
list_ = []
for file_ in allFiles:
    df = pd.read_csv(file_,encoding='latin-1')
    list_.append(df)
lieux = pd.concat(list_)

C:\Users\pc\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:2717: DtypeWarning: Columns (2) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [14]:
lieux.head(2)

Num_Acc  catr voie   v1   v2  circ  nbv   pr    pr1  vosp  prof  plan  \
0  200500000001   3.0   41  0.0    B   2.0  2.0  1.0  430.0   0.0   1.0   1.0   
1  200500000002   2.0   41  0.0  NaN   0.0  2.0  0.0    0.0   1.0   1.0   1.0   

   lartpc  larrout  surf  infra  situ  env1  
0     0.0     63.0   1.0    0.0   1.0   0.0  
1     0.0    100.0   1.0    0.0   5.0   0.0

In [15]:
lieux.shape

(765576, 18)

#### Les données sur les vehicules :

In [16]:
vehicules_2005=pd.read_csv('dataset/vehicules/vehicules_2005.csv',encoding='latin-1')
vehicules_2005.shape

(149764, 9)

In [17]:
vehicules_2005.head(2)

Num_Acc  senc  catv  occutc  obs  obsm  choc  manv num_veh
0  200500000001     0     7       0    0     2     1     1     A01
1  200500000001     0     7       0    0     2     8    10     B02

In [18]:
vehicules_2005.dtypes

Num_Acc     int64
senc        int64
catv        int64
occutc      int64
obs         int64
obsm        int64
choc        int64
manv        int64
num_veh    object
dtype: object

In [19]:
path =r'dataset/vehicules' # use your path
allFiles = glob.glob(path + "/*.csv")
frame = pd.DataFrame()
list_ = []
for file_ in allFiles:
    df = pd.read_csv(file_,encoding='latin-1')
    list_.append(df)
vehicules = pd.concat(list_)

In [20]:
vehicules.shape

(1307042, 9)

In [21]:
vehicules.head(2)

Num_Acc  senc  catv  occutc  obs  obsm  choc  manv num_veh
0  200500000001   0.0     7       0  0.0   2.0   1.0   1.0     A01
1  200500000001   0.0     7       0  0.0   2.0   8.0  10.0     B02

In [22]:
vehicules.dtypes

Num_Acc      int64
senc       float64
catv         int64
occutc       int64
obs        float64
obsm       float64
choc       float64
manv       float64
num_veh     object
dtype: object

##### Save Data prepared to csv files

In [23]:
#vehicules :
#vehicules.to_csv('vehicules_2005-2016.csv')
#lieux
#lieux.to_csv('lieux_2005-2016.csv')
#usager :
#usagers.to_csv('usager_2005-2016.csv')
#caracteristiques :
#caracteristiques.to_csv('caracteristiques_2005-2016.csv')

### Nettoyage des données :

---
####  La table caracteristiques 
---

Vu que la base de données contient assez d'erreur de saisie, il s'avére que l'étape de nettoyage est primordiale avant de passer à d'autre étapes.

In [24]:
# rappel  sur la taille des données :
caracteristiques.shape
# on 16 variables ( ce qui correspond bien au document pdf) et 765 576 accidents.

(765576, 16)

##### Variables catégorielles :

###### gps :

In [25]:

print(len(caracteristiques.gps.unique()))
caracteristiques.gps.unique()
#On voit que gps a 8 valeurs uniques, alors que dans le documents pdf il n'y a que 5 ( M,A,G,R et Y)

9


array(['M', nan, '0', 'A', 'G', 'R', 'S', 'T', 'Y'], dtype=object)

nan, 0 et S se sont considérées comme des valeurs manquantes

In [26]:
caracteristiques[caracteristiques.gps=='S'].shape

(3, 16)

In [27]:
caracteristiques[caracteristiques.gps=='0'].shape

(9, 16)

In [28]:
caracteristiques.gps.isnull().sum()

432956

In [29]:
# puisque cette variable est utilisée pour localiser l'accident, on va remplacer '0' et'S' par nan.
caracteristiques.loc[:,['gps']]=caracteristiques.loc[:,['gps']].replace('0',np.nan)
caracteristiques.loc[:,['gps']]=caracteristiques.loc[:,['gps']].replace('S',np.nan)

In [30]:
#verification :
print(len(caracteristiques.gps.unique()))
caracteristiques.gps.unique()

7


array(['M', nan, 'A', 'G', 'R', 'T', 'Y'], dtype=object)

On garde nan pour l'instant. 

###### jour :

In [31]:
print(len(caracteristiques.jour.unique()))
caracteristiques.jour.unique()
# pas de soucis pour cette variable

31


array([12, 21,  4, 10, 28,  3, 18, 25, 29, 23, 11,  1, 30, 19,  9, 31, 15,
       13, 26,  2,  6,  8, 20, 16,  7,  5, 24, 14, 17, 22, 27], dtype=int64)

###### mois :

In [32]:
print(len(caracteristiques.mois.unique()))
caracteristiques.mois.unique()
# pas de soucis pour cette variable


12


array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12], dtype=int64)

##### an :

In [33]:
print(len(caracteristiques.an.unique()))
caracteristiques.an.unique()
#pareil

11


array([ 5,  6,  7,  8, 10, 11, 12, 13, 14, 15, 16], dtype=int64)

###### lum :

In [34]:
print(len(caracteristiques.lum.unique()))
caracteristiques.lum.unique()
# bon

5


array([3, 1, 5, 4, 2], dtype=int64)

###### agg


In [35]:
print(len(caracteristiques['agg'].unique()))
caracteristiques['agg'].unique()
#bon

2


array([2, 1], dtype=int64)

###### intersection :

In [36]:
print(len(caracteristiques.int.unique()))
caracteristiques.int.unique()

10


array([1, 2, 9, 0, 6, 8, 3, 4, 7, 5], dtype=int64)

In [37]:
# 0 est un intrus 
caracteristiques[caracteristiques.int==0].shape
# il y 106 valeurs de 0
# on va remplacer cette valeur par la valeur la plus courante 

(106, 16)

In [38]:
caracteristiques.int.astype(str).describe()
# le plus courant c'est 1 avec une fréquence de 547440
# c'est normale car la majorité des accidents n'ont pas lieu sur les intersections ( vitesse très faible générelement)

count     765576
unique        10
top            1
freq      547440
Name: int, dtype: object

In [39]:
# remplacer 0 par 1: 
caracteristiques.loc[:,['int']]=caracteristiques.loc[:,['int']].replace(0,1)

In [40]:
#verifications :
print(len(caracteristiques.int.unique()))
caracteristiques.int.unique()

9


array([1, 2, 9, 6, 8, 3, 4, 7, 5], dtype=int64)

###### atm :

In [41]:
print(len(caracteristiques.atm.unique()))
caracteristiques.atm.unique()
# il y des nan values, de même on va les remplacer par la valeur la plus courante

10


array([  1.,   2.,   3.,   8.,   9.,   7.,   6.,   4.,   5.,  nan])

In [42]:
# taille des nan values dans atm :
print('taille :', caracteristiques.atm.isnull().sum())
# valeur la plus courante dans atm :
caracteristiques.atm.astype(str).describe()
# 1 est la valeur la plus courante ( normale)

taille : 48


count     765576
unique        10
top          1.0
freq      619775
Name: atm, dtype: object

In [43]:
#remplacer nan par 1 :
caracteristiques.loc[:,['atm']]=caracteristiques.loc[:,['atm']].replace(np.nan,1)

In [44]:
#verification :
print(len(caracteristiques.atm.unique()))
caracteristiques.atm.unique()

9


array([ 1.,  2.,  3.,  8.,  9.,  7.,  6.,  4.,  5.])

###### collision :

In [45]:
print(len(caracteristiques.col.unique()))
caracteristiques.col.unique()
# nan values are in 

8


array([  3.,   1.,   5.,   6.,   2.,   4.,   7.,  nan])

In [46]:
# taille des nan values dans col :
print('taille :', caracteristiques.col.isnull().sum())
# valeur la plus courante dans col :
caracteristiques.col.astype(str).describe()
# 6 est la valeur la plus courante ( autre collision)

taille : 11


count     765576
unique         8
top          6.0
freq      252916
Name: col, dtype: object

In [47]:
#remplacer nan par 6 :
caracteristiques.loc[:,['col']]=caracteristiques.loc[:,['col']].replace(np.nan,6)

In [48]:
#verification :
print(len(caracteristiques.col.unique()))
caracteristiques.col.unique()

7


array([ 3.,  1.,  5.,  6.,  2.,  4.,  7.])

###### Adresse postale :

In [49]:
print(len(caracteristiques.adr.unique()))
caracteristiques.adr.unique()
# des adresse, donc pas de problémes ( sauf erreur de saisie. on va lire adresse par adresse bien sûr pour s'assurer )

337657


array(['CD41B', 'rue de Lille', nan, ..., 'tsoundzou-KOUALE ( rue d',
       "passamanty-M'hogoni (rue", 'kaweni-GEOLE (la)'], dtype=object)

##### Variables numériques :

In [50]:
# valeur nan par variable
caracteristiques.isnull().sum()

Num_Acc         0
an              0
mois            0
jour            0
hrmn            0
lum             0
agg             0
int             0
atm             0
col             0
com             2
adr        127664
gps        432968
lat        436603
long       436607
dep             0
dtype: int64

###### lon/lat

 lat et lon sont des variables numériques. Ils présentent plus de 50% de valeurs manquantes. On les garde comme ça pour l'instant.

###### com :

Chaque dépertement contient des communes. Dans notre base les déparetements sont chiffrés entre 10 et 976. 

RQ :
    - on peut aggréger nos données par département et étudier chaque département tout seul
    - On peut encore ajouter de la granulité et aggréger par commune.


In [51]:
#nan ==> le mode
caracteristiques.com.astype(str).describe()

count     765576
unique       921
top         55.0
freq       31375
Name: com, dtype: object

In [52]:
caracteristiques.loc[:,['com']]=caracteristiques.loc[:,['com']].replace(np.nan,55)

###### dep

In [53]:
#pas de valeur manquante pour cette variables
print("il y a ",len(caracteristiques.dep.unique()),"départements ")
print(sorted(list(caracteristiques.dep.unique())))


il y a  101 départements 
[10, 20, 30, 40, 50, 60, 70, 80, 90, 100, 110, 120, 130, 140, 150, 160, 170, 180, 190, 201, 202, 210, 220, 230, 240, 250, 260, 270, 280, 290, 300, 310, 320, 330, 340, 350, 360, 370, 380, 390, 400, 410, 420, 430, 440, 450, 460, 470, 480, 490, 500, 510, 520, 530, 540, 550, 560, 570, 580, 590, 600, 610, 620, 630, 640, 650, 660, 670, 680, 690, 700, 710, 720, 730, 740, 750, 760, 770, 780, 790, 800, 810, 820, 830, 840, 850, 860, 870, 880, 890, 900, 910, 920, 930, 940, 950, 971, 972, 973, 974, 976]


In [54]:
dep_values=list(caracteristiques.dep.astype(str).values)

In [55]:
set([dep for dep in dep_values if len(dep)==2])
# on a bien toutes les valeurs correspondantes au dépertement de numéro entre 1 et 9.

{'10', '20', '30', '40', '50', '60', '70', '80', '90'}

In [56]:
print(set([dep for dep in dep_values if len(dep)==3 and dep[-1]!='0' and dep!='202' and dep!='201'])) 

{'971', '976', '972', '973', '974'}


In [57]:
# 201 et 202 remplacent 2A et 2B ( voir lien www.francegene.com/rech-fr/dep-fr.php) : donc pour ces deux valeurs pas de soucis
# mais les autres ils ne se terminent pas par 0 ?
#  normalement il y a 95 departement  d'après le lien ci-dessous, donc 950 doit être la plus grande valeur.
# A voir : soit se sont des nouveau dépertement ? sinon on va les remplacer par la plus courante valeurs.
# en fait les autres numéros sont celles de la france d'outre mer. plus précisement :
# 971 : num départemental de Guadeloupe
# 972 : Maritinique
# 973 : La Guyane 
# 974 : La Réunion
# 976 : Mayotte
# Donc c'est à nous de choisir : est ce que juste la france europééne ou aussi la france d'outre mer.

In [58]:
#save table to csv file :
caracteristiques.to_csv('caracteristiques_2005-2016.csv',index=False,)

---
#### la table lieux : 
---

##### variables catégorielles :

In [59]:
# shape :
lieux.shape

(765576, 18)

###### catégorie de route :

In [60]:
print(len(lieux.catr.unique()))
lieux.catr.unique()
# il y des nan values

8


array([  3.,   2.,   4.,   6.,   9.,   5.,   1.,  nan])

In [61]:
# 9 spécifie "autre" de type de route
print(lieux[lieux.catr==9].shape)  
#taille valeur nan
print(lieux.catr.isnull().sum())

(14992, 18)
1


In [62]:
# remplacer nan par 9.( 9 toujours n'est pas connu donc implicitement c'est nan)
lieux.loc[:,['catr']]=lieux.loc[:,['catr']].replace(np.nan,9)

In [63]:
#verification
lieux.catr.isnull().sum()

0

###### régime circulation :

In [64]:
print(len(lieux.circ.unique()))
lieux.circ.unique()
# il y a 0 et nan qui ne sont pas des valeurs pour cette variables

6


array([  2.,   0.,   3.,   4.,   1.,  nan])

In [65]:
print("taille 0",lieux[lieux.circ==0].shape)
print("taille nan",lieux.circ.isnull().sum())

taille 0 (40732, 18)
taille nan 737


In [66]:
lieux.circ.astype(str).describe()
# top est 2( les accidents ont lieu beaucoup sur les  routes nationals)

count     765576
unique         6
top          2.0
freq      485032
Name: circ, dtype: object

In [67]:
# On va remplacer ces deux variables par la plus courante valeur 
lieux.loc[:,['circ']]=lieux.loc[:,['circ']].replace(np.nan,2)
lieux.loc[:,['circ']]=lieux.loc[:,['circ']].replace(0,2)

In [68]:
# verification :
lieux.circ.unique()

array([ 2.,  3.,  4.,  1.])

###### voie réservée :

In [69]:
print(len(lieux.vosp.unique()))
lieux.vosp.unique()
# 0 et nan doivent être changés

5


array([  0.,   1.,   3.,   2.,  nan])

In [70]:
print("taille 0 :",lieux[lieux.vosp==0].shape)
print("taille nan :",lieux.vosp.isnull().sum())
print(lieux.shape)
# on remarque que quasiment toute la colonne est en '0'.
# on remplace nan par 0 et on laisse cette variable.

taille 0 : (717337, 18)
taille nan : 1382
(765576, 18)


In [71]:
lieux.loc[:,['vosp']]=lieux.loc[:,['vosp']].replace(0,np.nan)


###### profil de la route : 

In [72]:
print(len(lieux.prof.unique()))
lieux.prof.unique()

6


array([  1.,   0.,   2.,   3.,   4.,  nan])

In [73]:
print("taille 0 :",lieux[lieux.prof==0].shape)
print("taille nan :",lieux.prof.isnull().sum())
print(lieux.shape)

taille 0 : (58716, 18)
taille nan : 982
(765576, 18)


In [74]:
# on remplace 0 et nan par le mode :
lieux.prof.astype(str).describe()
# top est 1 (Plat)

count     765576
unique         6
top          1.0
freq      578523
Name: prof, dtype: object

In [75]:
lieux.loc[:,['prof']]=lieux.loc[:,['prof']].replace(0,1)
lieux.loc[:,['prof']]=lieux.loc[:,['prof']].replace(np.nan,1)

###### Tracé en plan :
    
    

In [76]:
print(len(lieux.plan.unique()))
lieux.plan.unique()

6


array([  1.,   3.,   2.,   0.,   4.,  nan])

In [77]:
print("taille 0 :",lieux[lieux.plan==0].shape)
print("taille nan :",lieux.plan.isnull().sum())
print(lieux.shape)

taille 0 : (52708, 18)
taille nan : 1001
(765576, 18)


In [78]:
# on remplace 0 et nan par le mode :
lieux.plan.astype(str).describe()
# top est 1 ( Partie rectiligne)

count     765576
unique         6
top          1.0
freq      583560
Name: plan, dtype: object

In [79]:
lieux.loc[:,['plan']]=lieux.loc[:,['plan']].replace(0,1)
lieux.loc[:,['plan']]=lieux.loc[:,['plan']].replace(np.nan,1)

###### Etat de la surface 

In [80]:
print(len(lieux.surf.unique()))
lieux.surf.unique()

11


array([  1.,   2.,   0.,   9.,   7.,   8.,   5.,   6.,   3.,   4.,  nan])

In [81]:
print("taille 0 :",lieux[lieux.surf==0].shape)
print("taille nan :",lieux.surf.isnull().sum())
print(lieux.shape)

taille 0 : (24885, 18)
taille nan : 949
(765576, 18)


In [82]:
# on remplace par 9 (autre: 
lieux.loc[:,['surf']]=lieux.loc[:,['surf']].replace(0,9)
lieux.loc[:,['surf']]=lieux.loc[:,['surf']].replace(np.nan,9)

###### situation de l'acccident

In [83]:
print(len(lieux.situ.unique()))
lieux.situ.unique()

7


array([  1.,   5.,   3.,   4.,   0.,   2.,  nan])

In [84]:
print("taille 0 :",lieux[lieux.situ==0].shape)
print("taille nan :",lieux.situ.isnull().sum())
print(lieux.shape)

taille 0 : (39296, 18)
taille nan : 943
(765576, 18)


In [85]:
# on remplace par le mode :
lieux.situ.astype(str).describe()
#top est 1 ( sur chaussé)

count     765576
unique         7
top          1.0
freq      658335
Name: situ, dtype: object

In [86]:

lieux.loc[:,['situ']]=lieux.loc[:,['situ']].replace(0,1)
lieux.loc[:,['situ']]=lieux.loc[:,['situ']].replace(np.nan,1)

###### point école :

In [87]:
# on sait pas pour cette variables quelles sont les veleurs prises
lieux.env1.unique()
# à voir ...


array([  0.,   3.,  99.,  nan])

##### variables numériques :

In [88]:
lieux.dtypes
# le reste des variables sont des floats réellement ( donc on voit si y des des nan on remplace par mean)

Num_Acc      int64
catr       float64
voie        object
v1         float64
v2          object
circ       float64
nbv        float64
pr         float64
pr1        float64
vosp       float64
prof       float64
plan       float64
lartpc     float64
larrout    float64
surf       float64
infra      float64
situ       float64
env1       float64
dtype: object

In [89]:
lieux.isnull().sum()

Num_Acc         0
catr            0
voie        55754
v1         433094
v2         734917
circ            0
nbv          1148
pr         397330
pr1        398329
vosp       718719
prof            0
plan            0
lartpc       9005
larrout      7920
surf            0
infra        1110
situ            0
env1         1127
dtype: int64

###### voie :

In [90]:
lieux.voie.describe()
# 55754 nan
# contient beaucoup de 0 alors que 0 n'est pas une numéro de route 
# on transforme 0 à nan puis on voit le mode puis (nan==> le mode)

count     709822.0
unique      2469.0
top            0.0
freq      399947.0
Name: voie, dtype: float64

In [91]:
lieux.loc[:,['voie']]=lieux.loc[:,['voie']].replace(0,np.nan)


In [92]:
lieux.voie.describe()

count     309875.0
unique      2468.0
top            1.0
freq       19374.0
Name: voie, dtype: float64

In [93]:
lieux.loc[:,['voie']]=lieux.loc[:,['voie']].replace(np.nan,1)

###### Indice numériques du numéro de la route : 

In [94]:
lieux.v1.unique()

array([  0.,   2.,   1.,   9.,   6.,   4.,   5.,   7.,   3.,   8.,  nan])

In [95]:
lieux.v1.astype(str).describe() 
# à voir comment régler cette variable

count     765576
unique        11
top          nan
freq      433094
Name: v1, dtype: object

###### Lettre V2

In [96]:
lieux.v2.unique()
# y a une probléme de encoding : /x04
# y des nan 
# y des 0  
# il faut comprendre la variable pour 

array(['B', nan, 'A', 'E', 'N', 'D', 'C', 'F', 'R', 'G', 'J', 'b', 'a',
       'S', 'W', 'H', 'P', '\x04', 'K', 'T', 'L', 'M', 'Z', 'O', 'I', 's',
       'n', 'Q', 'e', 'V', 'd', 'c', 'z', 'U', 'v', 'm', 'Y', 'w', 'y',
       't', '0', '4', '1', '8', '2', '5', '6', '9', '3', '7', 'f', 'p',
       'g', 'l', 'r', 'X'], dtype=object)

###### nombre total de voie de circulation  :

In [97]:
print(sorted(set(list(lieux.nbv.unique()))))

[0.0, 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0, 11.0, 12.0, 13.0, 14.0, 15.0, 16.0, 17.0, 18.0, 20.0, 21.0, 22.0, 23.0, 24.0, 25.0, 26.0, 27.0, 28.0, 29.0, 30.0, 31.0, nan, 33.0, 36.0, 39.0, 40.0, 41.0, 42.0, 44.0, 45.0, 50.0, 52.0, 53.0, 54.0, 60.0, 62.0, 65.0, 70.0, 76.0, 84.0, 86.0, 90.0, 91.0, 99.0]


In [98]:
# on remplace nan par la moyenne ( méthode stationnaire d'imputation dans le cas des MCAR)
lieux.loc[:,['nbv']]=lieux.loc[:,['nbv']].replace(np.nan,int(lieux.nbv.mean())) #  on prends la partie entiére

###### Numéro PR :
    

In [99]:
lieux.loc[:,['pr']]=lieux.loc[:,['pr']].replace(np.nan,int(lieux.pr.mean())) 

###### Distance en mètres au PR :

In [100]:
lieux.loc[:,['pr1']]=lieux.loc[:,['pr1']].replace(np.nan,lieux.pr1.mean()) 

###### Largeur du terre plein central TPC ( s il exist )

In [101]:

lieux.loc[:,['lartpc']]=lieux.loc[:,['lartpc']].replace(np.nan,lieux.lartpc.mean()) 

###### larrout

In [102]:
lieux.loc[:,['larrout']]=lieux.loc[:,['larrout']].replace(np.nan,lieux.larrout.mean())

In [103]:
# save table to csv file
lieux.to_csv('lieux_2005-2016.csv',index=False)

---
#### la table usagers 
---

In [104]:
#shape :
usagers.shape

(1710043, 12)

In [105]:
print(usagers.dtypes)

Num_Acc      int64
place      float64
catu         int64
grav         int64
sexe         int64
trajet     float64
secu       float64
locp       float64
actp       float64
etatp      float64
an_nais    float64
num_veh     object
dtype: object


###### place : 

In [106]:
usagers.place.unique()

array([  1.,   2.,   4.,   5.,   3.,   0.,   8.,   9.,   7.,   6.,  nan])

In [107]:
print("taille 0 :" , usagers[usagers.place==0].shape)
print("taille nan :" , usagers.place.isnull().sum())


taille 0 : (60766, 12)
taille nan : 86734


In [108]:
#on les remplace par ? 
# ici les valeurs manquantes vont dependre de la nature du vehicule 
# donc pour chaque type de vehicule on va remplacer les valeurs manquates par le mode.
# donc on verra cette variable apès avoir fait la jointure entre les tables 
#sinon on peut simplifier et remplacer par le mode:
usagers.place.astype(str).describe()

count     1710043
unique         11
top           1.0
freq      1275547
Name: place, dtype: object

In [109]:
usagers.loc[:,['place']]=usagers.loc[:,['place']].replace(0,1)
usagers.loc[:,['place']]=usagers.loc[:,['place']].replace(np.nan,1)

###### Catégorie d'usager :

In [110]:
usagers.catu.unique()
#c'est bon

array([1, 2, 3, 4], dtype=int64)

###### Gravité de l'accident​ :

In [111]:
usagers.grav.unique()
#c'est bon

array([4, 3, 1, 2], dtype=int64)

###### Sexe de l'usager :


In [112]:
usagers.sexe.unique()
#c'est bon

array([1, 2], dtype=int64)

###### Annee de naissance de l'usager'

In [113]:
usagers.an_nais.dtype
# pas de soucis ici sauf si il y une erreur dans la saisie  d'une data :
print('min des dates :',usagers.an_nais.min())
print('max des dates :',usagers.an_nais.max())

min des dates : 1896.0
max des dates : 2016.0


Si les accidents ont lieu entre 2005-2016, alors un passer né en 1896 il a eu à la date d'accident en minimum un âge de 109 
et au max 120 ans. Ces deux date restent acceptable ~~

###### Motif du déplacement au moment de l’accident :

In [114]:
usagers.trajet.unique()

array([  1.,   3.,   0.,   5.,   9.,   4.,   2.,  nan])

In [115]:
print("taille 0 :" , usagers[usagers.trajet==0].shape)
print("taille nan :" , usagers.trajet.isnull().sum())

taille 0 : (506839, 12)
taille nan : 272


In [116]:
# on remplace par 9 :
usagers.loc[:,['trajet']]=usagers.loc[:,['trajet']].replace(0,9)
usagers.loc[:,['trajet']]=usagers.loc[:,['trajet']].replace(np.nan,9)

###### secu :

In [117]:
usagers.secu.unique()

array([ 11.,  31.,  21.,   0.,  12.,  13.,  90.,  10.,  22.,  91.,  33.,
        41.,  23.,  20.,  92.,  32.,  43.,  93.,  42.,   1.,  nan,  30.,
        40.,   3.,   2.])

In [118]:
secu_values=list(usagers.secu.astype(str).values)


In [119]:
# on traite d'abord le len puis les nan :
secu_values_set=[secu[:-2] for secu in secu_values if secu!='nan']
# d'après le document chaque valeur est censée d'avoir deux chiffres:

In [120]:
print(set(secu_values_set))
# il y le 0 et parfois on a qu'un seul chiffre ( ex 2, 1, 3)

{'1', '30', '22', '92', '0', '90', '43', '91', '41', '13', '23', '12', '42', '2', '31', '3', '32', '40', '20', '93', '33', '11', '10', '21'}


In [121]:
usagers.secu.astype(str).describe()
#top 11 : ceinture exist et elle est utilisée

count     1710043
unique         25
top          11.0
freq       952293
Name: secu, dtype: object

In [122]:
# remplacer nan et 0 par 11 :
usagers.loc[:,['secu']]=usagers.loc[:,['secu']].replace(0,11)
usagers.loc[:,['secu']]=usagers.loc[:,['secu']].replace(np.nan,11)

In [123]:
# pour le reste : soit on simplifie et on remplace par 11
# soit on compléte le chiffre par un deuxième qui soit le mode approprié au chiffre existant :
# deuxième méthode est plus logique :
from statistics import mode
secu_1=[secu for secu in secu_values if '1' in secu]
mode(secu_1)
# donc on remplace par 11


'11.0'

In [124]:
secu_2=[secu for secu in secu_values if '2' in secu]
mode(secu_2)
# on remplace par '21'

'21.0'

In [125]:
secu_3=[secu for secu in secu_values if '3' in secu]
mode(secu_3)
# on remplace par '13'

'13.0'

In [126]:
usagers.loc[:,['secu']]=usagers.loc[:,['secu']].replace(1,11)
usagers.loc[:,['secu']]=usagers.loc[:,['secu']].replace(2,21)
usagers.loc[:,['secu']]=usagers.loc[:,['secu']].replace(3,13)


###### Localisation du piéton :

In [127]:
usagers.locp.unique()

array([  0.,   2.,   4.,   1.,   5.,   6.,   3.,   8.,   7.,  nan])

In [128]:
print("taille 0 :" , usagers[usagers.locp==0].shape)
print("taille nan :" , usagers.locp.isnull().sum())
print(usagers.shape)

taille 0 : (1579313, 12)
taille nan : 1071
(1710043, 12)


In [129]:
# ici on va laisser 0 comme une autre valeur, il se peut qu'elle aie pour signification : 'il n'y a pas de piéton'
# on remplace nan par 0 également 
usagers.loc[:,['locp']]=usagers.loc[:,['locp']].replace(np.nan,0)

###### Action du piéton :

In [130]:
usagers.actp.unique()

array([  0.,   3.,   1.,   5.,   2.,   9.,   6.,   4.,  nan])

In [131]:
# nan ==> 0
usagers.loc[:,['locp']]=usagers.loc[:,['locp']].replace(np.nan,0)

###### etatp

In [132]:
usagers.etatp.unique()

array([  0.,   2.,   1.,   3.,  nan])

In [133]:
# on laisse 0 comme une autre valeur qui va indiquer qu'il n'y avait pas de piéton
# nan ==> 0
usagers.loc[:,['etatp']]=usagers.loc[:,['etatp']].replace(np.nan,0)

Save to csv file :

In [134]:
usagers.to_csv('usager_2005-2016.csv', index=False)

---
la table des vehicules
---

In [135]:
vehicules.shape

(1307042, 9)

In [136]:
vehicules.dtypes


Num_Acc      int64
senc       float64
catv         int64
occutc       int64
obs        float64
obsm       float64
choc       float64
manv       float64
num_veh     object
dtype: object

###### Sens de circulation

In [137]:
vehicules.senc.unique()

array([  0.,   1.,   2.,  nan])

In [138]:
print("taille 0 :" , vehicules[vehicules.senc==0].shape)
print("taille nan :" , vehicules.senc.isnull().sum())

taille 0 : (1199403, 9)
taille nan : 72


In [139]:
vehicules.senc.astype(str).describe()

count     1307042
unique          4
top           0.0
freq      1199403
Name: senc, dtype: object

In [140]:
# on remplace par une valeur au choix :  sens croissant 
vehicules.loc[:,['senc']]=vehicules.loc[:,['senc']].replace(np.nan,1)
vehicules.loc[:,['senc']]=vehicules.loc[:,['senc']].replace(0,1)

###### Catégorie du véhicule :

In [141]:
print(len(vehicules.catv.unique()))
vehicules.catv.unique()
#Ok

33


array([ 7,  2,  9,  5, 18, 17,  1, 10,  4, 15, 19, 14, 13, 12, 21, 20, 99,
        3, 16,  8,  6, 11, 38, 32, 37, 30, 36, 31, 33, 39, 34, 35, 40], dtype=int64)

###### Obstacle mobile heurté :

In [142]:
vehicules.obsm.unique()

array([  2.,   0.,   1.,   9.,   4.,   5.,   6.,  nan])

In [143]:
# nan,0===>9
vehicules.loc[:,['obsm']]=vehicules.loc[:,['obsm']].replace(np.nan,9)
vehicules.loc[:,['obsm']]=vehicules.loc[:,['obsm']].replace(0,9)

###### Point de choc initial :

In [144]:
vehicules.choc.unique()

array([  1.,   8.,   7.,   0.,   3.,   2.,   6.,   4.,   5.,   9.,  nan])

In [145]:
# 0, nan ==> mode :
vehicules.choc.astype(str).describe()

count     1307042
unique         11
top           1.0
freq       485214
Name: choc, dtype: object

In [146]:
vehicules.loc[:,['choc']]=vehicules.loc[:,['choc']].replace(np.nan,1)
vehicules.loc[:,['choc']]=vehicules.loc[:,['choc']].replace(0,1)

###### Nombre d’occupants dans le transport en commun

In [147]:
vehicules.occutc.unique()
# 0 pour dire que ce n'est pas transport commun.

array([  0,   1,   2,  26,  23,  36,  31,   6,  51,  27,  52,  25,  10,
         8,  14,  20,  11,  40,  33,  12,   3,  13,  45,  39,   5,  15,
        32, 100,  30,  29,  35,  41,   4,  21,  16,  18,  37,  60,  62,
        46,  50,  28,  44,   7,  57,  42,   9,  17,  58,  24,  43,  48,
        22, 470,  69,  61, 150,  47, 120,  66,  49,  38,  54,  59, 200,
        34,  53,  55,  96,  65, 210,  63,  93,  80,  85, 213,  64, 193,
       111,  70, 700, 300,  73,  74,  19,  72, 110,  97,  56,  75, 220,
        81,  95,  84,  83,  67,  76,  89,  90, 480, 140,  99, 240,  71,
       132, 125, 153, 184, 170, 130, 490, 250], dtype=int64)

###### manv

In [148]:
# manv de type int
vehicules.manv.isnull().sum()

227

In [149]:
vehicules.manv.astype(str).describe()

count     1307042
unique         26
top           1.0
freq       575354
Name: manv, dtype: object

In [150]:
vehicules.loc[:,['manv']]=vehicules.loc[:,['manv']].replace(np.nan,1)

save to csv file 

In [151]:
vehicules.to_csv('vehicules_2005-2016.csv', index=False)

---
Fin nettoyage 
---

### La jointure des tables :

In [152]:
vehicules.head(5)

Num_Acc  senc  catv  occutc  obs  obsm  choc  manv num_veh
0  200500000001   1.0     7       0  0.0   2.0   1.0   1.0     A01
1  200500000001   1.0     7       0  0.0   2.0   8.0  10.0     B02
2  200500000002   1.0     7       0  0.0   2.0   7.0  16.0     A01
3  200500000002   1.0     2       0  0.0   2.0   1.0   1.0     B02
4  200500000003   1.0     2       0  0.0   2.0   1.0   1.0     A01

In [153]:
usagers.head(6)

Num_Acc  place  catu  grav  sexe  trajet  secu  locp  actp  etatp  \
0  200500000001    1.0     1     4     1     1.0  11.0   0.0   0.0    0.0   
1  200500000001    1.0     1     3     2     3.0  11.0   0.0   0.0    0.0   
2  200500000001    2.0     2     1     1     9.0  11.0   0.0   0.0    0.0   
3  200500000001    4.0     2     1     1     9.0  31.0   0.0   0.0    0.0   
4  200500000001    5.0     2     1     1     9.0  11.0   0.0   0.0    0.0   
5  200500000001    3.0     2     1     2     9.0  11.0   0.0   0.0    0.0   

   an_nais num_veh  
0   1976.0     A01  
1   1968.0     B02  
2   1964.0     B02  
3   2004.0     B02  
4   1998.0     B02  
5   1991.0     B02

In [154]:
print(lieux.shape)
print(caracteristiques.shape) 

(765576, 18)
(765576, 16)


on commence par ces deux table. PK/FK est Num_Acc

In [155]:
# je choisi que 90 000 ligne pour le test ( et d'ailler ca passe pas la taille entière sur mon pc)
carac_lieux_marged=pd.merge(caracteristiques.iloc[:90000,:],lieux.iloc[:90000,:],on='Num_Acc')

In [156]:
carac_lieux_marged.head()

Num_Acc  an  mois  jour  hrmn  lum  agg  int  atm  col ...     pr1  \
0  200500000001   5     1    12  1900    3    2    1  1.0  3.0 ...   430.0   
1  200500000002   5     1    21  1600    1    2    1  1.0  1.0 ...     0.0   
2  200500000003   5     1    21  1845    3    1    1  2.0  1.0 ...     0.0   
3  200500000004   5     1     4  1615    1    1    1  1.0  5.0 ...     0.0   
4  200500000005   5     1    10  1945    3    1    1  3.0  6.0 ...   630.0   

  vosp prof  plan  lartpc  larrout  surf infra  situ env1  
0  NaN  1.0   1.0     0.0     63.0   1.0   0.0   1.0  0.0  
1  1.0  1.0   1.0     0.0    100.0   1.0   0.0   5.0  0.0  
2  1.0  1.0   1.0     0.0      0.0   2.0   0.0   5.0  0.0  
3  NaN  1.0   1.0     0.0      0.0   1.0   0.0   1.0  0.0  
4  NaN  1.0   3.0     0.0     59.0   2.0   0.0   3.0  0.0  

[5 rows x 33 columns]

on merge avec vehicules : 

In [157]:
vehicules.head()

Num_Acc  senc  catv  occutc  obs  obsm  choc  manv num_veh
0  200500000001   1.0     7       0  0.0   2.0   1.0   1.0     A01
1  200500000001   1.0     7       0  0.0   2.0   8.0  10.0     B02
2  200500000002   1.0     7       0  0.0   2.0   7.0  16.0     A01
3  200500000002   1.0     2       0  0.0   2.0   1.0   1.0     B02
4  200500000003   1.0     2       0  0.0   2.0   1.0   1.0     A01

In [158]:
carac_lieux_veh_merged=pd.merge(carac_lieux_marged, vehicules,on='Num_Acc' )

In [159]:
carac_lieux_veh_merged.head()

Num_Acc  an  mois  jour  hrmn  lum  agg  int  atm  col   ...    situ  \
0  200500000001   5     1    12  1900    3    2    1  1.0  3.0   ...     1.0   
1  200500000001   5     1    12  1900    3    2    1  1.0  3.0   ...     1.0   
2  200500000002   5     1    21  1600    1    2    1  1.0  1.0   ...     5.0   
3  200500000002   5     1    21  1600    1    2    1  1.0  1.0   ...     5.0   
4  200500000003   5     1    21  1845    3    1    1  2.0  1.0   ...     5.0   

  env1 senc  catv  occutc  obs  obsm choc  manv num_veh  
0  0.0  1.0     7       0  0.0   2.0  1.0   1.0     A01  
1  0.0  1.0     7       0  0.0   2.0  8.0  10.0     B02  
2  0.0  1.0     7       0  0.0   2.0  7.0  16.0     A01  
3  0.0  1.0     2       0  0.0   2.0  1.0   1.0     B02  
4  0.0  1.0     2       0  0.0   2.0  1.0   1.0     A01  

[5 rows x 41 columns]

save to csv file 

In [160]:
carac_lieux_veh_merged.to_csv("carac_lieux_veh_merged.csv", index=False)

In [161]:
usagers.shape

(1710043, 12)

In [162]:
usagers.head()

Num_Acc  place  catu  grav  sexe  trajet  secu  locp  actp  etatp  \
0  200500000001    1.0     1     4     1     1.0  11.0   0.0   0.0    0.0   
1  200500000001    1.0     1     3     2     3.0  11.0   0.0   0.0    0.0   
2  200500000001    2.0     2     1     1     9.0  11.0   0.0   0.0    0.0   
3  200500000001    4.0     2     1     1     9.0  31.0   0.0   0.0    0.0   
4  200500000001    5.0     2     1     1     9.0  11.0   0.0   0.0    0.0   

   an_nais num_veh  
0   1976.0     A01  
1   1968.0     B02  
2   1964.0     B02  
3   2004.0     B02  
4   1998.0     B02

on merge finalement avec usagers :

In [165]:
#  later 
#final_data=pd.merge(carac_lieux_veh_merged,usagers,on=['num_veh','num_veh'])